// @kui-shell/plugin-kubeui/mdist/index.js
export { isKubeResource, isCrudableKubeResource, isPod, isDeployment, isKubeItems, isJob } from './lib/model/resource';
export { default as TrafficLight } from './lib/model/traffic-light';
export { default as apiVersion } from './controller/kubectl/apiVersion';
export { doExecWithPty, doExecWithStdout, doExecWithStatus, doExecWithTable } from './controller/kubectl/exec';
export { doExecRaw, doNativeExec } from './controller/kubectl/raw';
export { default as commandPrefix } from './controller/command-prefix';
export { default as defaultFlags } from './controller/kubectl/flags';
export { hasLabel, getLabel, getLabelForArgv, isHelpRequest, getNamespace, getNamespaceForArgv, isForAllNamespaces } from './controller/kubectl/options';
export { default as parseName } from './lib/util/name';
export { formatTable, preprocessTable } from './lib/view/formatTable';
export { renderHelp } from './lib/util/help';
export { fetchFileString } from './lib/util/fetch-file';
export { fqnOf, fqn } from './controller/kubectl/fqn';
export { doGet } from './controller/kubectl/get';
export { doRun } from './controller/kubectl/run';
export { doCreate } from './controller/kubectl/create';
export { doDelete } from './controller/kubectl/delete';
export { doDescribe } from './controller/kubectl/describe';
