// @kui-shell/plugin-search/mdist/src-web/views/modes/yaml.js
"use strict";
/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
exports.__esModule = true;
var jsYaml = require("js-yaml");
var lodash = require("lodash");
/**
 * Render resources yaml tab
 *
 */
exports.yamlTab = function (resource) {
    var badges = [];
    // This will allow the sidecar balloon element to display the resources name.
    var balloon = resource.metadata.name.split(/(-[0-9])/);
    badges.push(balloon[0]);
    return {
        type: 'custom',
        // isEntity: true,
        content: jsYaml.safeDump(resource),
        contentType: 'yaml',
        kind: lodash.get(resource, 'kind', ''),
        name: lodash.get(resource, 'metadata.name', ''),
        packageName: lodash.get(resource, 'metadata.namespace', ''),
        badges: badges,
        modes: [
            {
                defaultMode: true,
                mode: 'yaml',
                direct: function () { return exports.yamlTab(resource); },
                leaveBottomStripeAlone: true,
                label: 'YAML'
            },
        ]
    };
};
