// @kui-shell/plugin-client-common/dist/components/Content/Table/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = renderTable;

var React = _interopRequireWildcard(require("react"));

var _core = require("@kui-shell/core");

var _ = require("../../../");

var _PaginatedTable = _interopRequireDefault(require("./PaginatedTable"));

var _LivePaginatedTable = _interopRequireDefault(require("./LivePaginatedTable"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
 * Copyright 2020 IBM Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function renderTable(tab, repl, response, paginate = 20, toolbars = false, asGrid = false, onRender) {
  if ((0, _core.isWatchable)(response)) {
    return React.createElement(_.KuiContext.Consumer, null, config => React.createElement(_LivePaginatedTable.default, {
      tab: tab,
      repl: repl,
      response: response,
      paginate: paginate,
      title: !config.disableTableTitle,
      toolbars: toolbars,
      asGrid: asGrid,
      onRender: onRender
    }));
  } else {
    return React.createElement(_.KuiContext.Consumer, null, config => {
      return React.createElement(_PaginatedTable.default, {
        tab: tab,
        repl: repl,
        response: response,
        paginate: paginate,
        title: !config.disableTableTitle,
        toolbars: toolbars,
        asGrid: asGrid
      });
    });
  }
}
