// @kui-shell/core/dist/webapp/cli.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const debug_1 = require("debug");
const debug = debug_1.default('webapp/cli');
debug('loading');
const status_1 = require("./status");
exports.setStatus = status_1.setStatus;
const cancel_1 = require("./cancel");
exports.doCancel = cancel_1.default;
const popup_core_1 = require("./popup-core");
exports.isPopup = popup_core_1.isPopup;
const listen_1 = require("./listen");
exports.listen = listen_1.listen;
const tab_1 = require("./tab");
exports.isTab = tab_1.isTab;
exports.getTabId = tab_1.getTabId;
exports.sameTab = tab_1.sameTab;
exports.getTabFromTarget = tab_1.getTabFromTarget;
exports.getCurrentTab = tab_1.getCurrentTab;
const prompt_1 = require("./prompt");
exports.getBlockOfPrompt = prompt_1.getBlockOfPrompt;
exports.getPrompt = prompt_1.getPrompt;
exports.getBottomPrompt = prompt_1.getBottomPrompt;
exports.getInitialPrompt = prompt_1.getInitialPrompt;
exports.getCurrentPrompt = prompt_1.getCurrentPrompt;
exports.getPromptFromTarget = prompt_1.getPromptFromTarget;
exports.getPromptLeft = prompt_1.getPromptLeft;
exports.getCurrentPromptLeft = prompt_1.getCurrentPromptLeft;
const block_1 = require("./block");
exports.installBlock = block_1.installBlock;
exports.getInitialBlock = block_1.getInitialBlock;
exports.getCurrentBlock = block_1.getCurrentBlock;
exports.getCurrentProcessingBlock = block_1.getCurrentProcessingBlock;
var print_1 = require("./print");
exports.ok = print_1.ok;
debug('finished loading modules');
