// @kui-shell/plugin-logui/dist/modes/previous.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@kui-shell/core");
const plugin_kubeui_1 = require("@kui-shell/plugin-kubeui");
const table_1 = require("../renderers/table");
const strings = core_1.i18n('plugin-logui');
function isLogs(resource) {
    return resource.kind === 'logs';
}
function renderPrevious(tab, resource) {
    return __awaiter(this, void 0, void 0, function* () {
        return table_1.formatAsTable(yield plugin_kubeui_1.doExecRaw(`${resource.originatingCommand} --previous`, {}, {}), resource.metadata);
    });
}
exports.default = {
    when: isLogs,
    mode: {
        mode: 'previous',
        label: strings('Previous'),
        content: renderPrevious,
        priority: 10
    }
};
