// @kui-shell/plugin-search/dist/src-web/controller/helpfiles/savedsearchhelp.js
"use strict";
/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
exports.__esModule = true;
/** how many options to show at a time, i.e. before scrolling */
var nRowsInViewport = 1;
/** list of related commands */
var all = [
    'search',
];
var allExcept = function (cmd) { return all.filter(function (_) { return _ !== cmd; }); };
/* the breadcrumb chain */
var parents = ['savedsearches'];
var header = {
    search: 'Search across managed clouds for kubernetes resources',
    savedsearches: 'Show saved searches of Kubernetes resources across managed clouds'
};
/**
 * Usage model for the savedsearches plugin
 *
 */
exports.toplevel = {
    breadcrumb: 'savedsearches',
    command: 'savedsearches',
    title: 'Show saved searches of Kubernetes resources across managed clouds',
    header: 'This command helps you retrieve saved searches.',
    example: 'savedsearches',
    nRowsInViewport: 1,
    available: [
        {
            command: 'savedsearches',
            docs: header.savedsearches,
            dir: true,
            commandPrefix: null
        },
    ],
    related: ['search']
};
