// @kui-shell/plugin-search/dist/src-web/controller/helpfiles/savedsearchhelp.js
"use strict";
/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
exports.__esModule = true;
var i18n_1 = require("../../util/i18n");
/**
 * Usage model for the savedsearches plugin
 *
 */
exports.toplevel = {
    breadcrumb: 'savedsearches',
    command: 'savedsearches',
    title: i18n_1["default"]('savedsearchhelp.title'),
    header: i18n_1["default"]('savedsearchhelp.header'),
    example: i18n_1["default"]('savedsearchhelp.example'),
    nRowsInViewport: 1,
    available: [
        {
            command: 'savedsearches',
            docs: i18n_1["default"]('savedsearchhelp.title'),
            dir: true,
            commandPrefix: null
        },
    ],
    alias: 'ss',
    related: ['search']
};
