// @kui-shell/plugin-kubeui/mdist/lib/view/modes/last-applied.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { i18n } from '@kui-shell/core';
const strings = i18n('plugin-kubeui');
function getLastAppliedRaw(resource) {
    const annotations = resource.metadata.annotations;
    return annotations !== undefined ? annotations['kubectl.kubernetes.io/last-applied-configuration'] : undefined;
}
function hasLastApplied(resource) {
    return getLastAppliedRaw(resource) !== undefined;
}
const renderLastApplied = (tab, resource) => __awaiter(void 0, void 0, void 0, function* () {
    const { safeDump } = yield import('js-yaml');
    return {
        content: safeDump(JSON.parse(getLastAppliedRaw(resource))),
        contentType: 'yaml'
    };
});
export default {
    when: hasLastApplied,
    mode: {
        mode: 'last applied',
        label: strings('lastApplied'),
        content: renderLastApplied
    }
};
