// @kui-shell/core/dist/core/settings.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
var userdata_1 = require("./userdata");
exports.clearPreference = userdata_1.clearPreference;
exports.getPreference = userdata_1.getPreference;
exports.setPreference = userdata_1.setPreference;
exports.userDataDir = userdata_1.userDataDir;
exports.inBottomInputMode = typeof document !== 'undefined' && document.body.classList.contains('kui--bottom-input');
function uiThemes() {
    return __awaiter(this, void 0, void 0, function* () {
        const { prescanModel } = yield Promise.resolve().then(() => require('../plugins/plugins'));
        return prescanModel().themeSets.filter(_ => _);
    });
}
exports.uiThemes = uiThemes;
