// @kui-shell/plugin-search/dist/src-web/controller/actionHandler.js
"use strict";
/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
// Hack to workaround build issues with Carbon dependencies
if (!window || !window.navigator || !window.navigator.userAgent) {
    Object.defineProperty(window, 'navigator', { value: { userAgent: 'node' }, writable: true });
    Object.defineProperty(document, 'getElementById', { value: function (val) { return document.querySelector('#' + val); }, writable: true });
}
var React = require("react");
var ReactDOM = require("react-dom");
var carbon_components_react_1 = require("carbon-components-react");
var HTTPClient_1 = require("./HTTPClient");
var i18n_1 = require("../util/i18n");
var search_queries_1 = require("../definitions/search-queries");
exports.notify = function (content) {
    var node = document.createElement('div');
    node.classList.add('bx--toast-notification-content');
    var toast = function () {
        return (React.createElement("div", { className: 'notification' },
            React.createElement(carbon_components_react_1.ToastNotification, { kind: !content.warning ? !content.message ? 'success' : 'error' : 'warning', title: !content.warning ? !content.message ? content : content.message : content.warning, caption: new Date().toLocaleTimeString(), timeout: 5000 })));
    };
    ReactDOM.render(React.createElement(toast), node);
    return node;
};
function deleteSavedSearch(args) {
    if (args.argv.length === 1) {
        return 'ERROR: Received wrong number of parameters.\nUSAGE: deleteSavedSearch <saved-search-name>';
    }
    var name = args.command.replace('deleteSavedSearch ', '');
    return new Promise(function (resolve, reject) {
        var warningToDelete = true;
        // Check if the record exist before trying to delete.
        HTTPClient_1["default"]('post', 'mcm', search_queries_1.SAVED_SEARCH_QUERY)
            .then(function (res) {
            res['warning'] = i18n_1["default"]('modal.save.warning', [name]);
            res.data.items.forEach(function (record) {
                if (record.name === name)
                    warningToDelete = false; // Record is available
            });
            warningToDelete
                ? resolve(exports.notify(res))
                : null;
            HTTPClient_1["default"]('post', 'mcm', search_queries_1.DELETE_QUERY(name))
                .then(function (res) {
                resolve(res.data.deleteQuery.userQueries
                    ? exports.notify(i18n_1["default"]('modal.deleted.save.success', [name]))
                    : exports.notify(res.errors[0]));
            });
        });
    });
}
function deleteResource(args) {
    if (args.argv.length !== 6) {
        return 'ERROR: Received wrong number of parameters.\nUSAGE: deleteResource <resource-name> <resource-namespace> <resource-kind> <resource-cluster> <resource-selfLink>';
    }
    return new Promise(function (resolve, reject) {
        // delete resource args = (name, namespace, kind, cluster, selfLink)
        HTTPClient_1["default"]('post', 'mcm', search_queries_1.DELETE_RESOURCE(args.argv[1], args.argv[2], args.argv[3], args.argv[4], args.argv[5]))
            .then(function (res) {
            resolve(res.errors
                ? exports.notify(res.errors[0])
                : exports.notify(i18n_1["default"]('modal.deleted.resource', [args.argv[1]])));
        });
    });
}
var deleteSavedSearchUsage = {
    command: 'deleteSavedSearch',
    strict: 'deleteSavedSearch',
    title: 'Deletes a previously saved search',
    header: 'Deletes a previously saved search',
    example: 'deleteSavedSearch <saved-search-name>',
    optional: [
        { name: 'name', positional: true },
    ]
};
var deleteResourceUsage = {
    command: 'deleteResource',
    strict: 'deleteResource',
    title: 'Deletes a cluster resource',
    header: 'Deletes a cluster resource',
    example: 'deleteResource <resource-name> <resource-namespace> <resource-kind> <resource-cluster> <resource-selfLink>'
};
/**
 * Here we register as a listener for commands
 */
exports["default"] = (function (commandTree) { return __awaiter(void 0, void 0, void 0, function () {
    var deleteSavedSearchOpts, deleteResourceOpts;
    return __generator(this, function (_a) {
        deleteSavedSearchOpts = { deleteSavedSearchUsage: deleteSavedSearchUsage, noAuthOk: true, inBrowserOk: true };
        deleteResourceOpts = { deleteResourceUsage: deleteResourceUsage, noAuthOk: true, inBrowserOk: true };
        commandTree.listen("/deleteSavedSearch", deleteSavedSearch, deleteSavedSearchOpts);
        commandTree.listen("/deleteResource", deleteResource, deleteResourceOpts);
        return [2 /*return*/];
    });
}); });
