// @kui-shell/plugin-kubeui/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var resource_1 = require("./lib/model/resource");
exports.isKubeResource = resource_1.isKubeResource;
exports.isCrudableKubeResource = resource_1.isCrudableKubeResource;
exports.isPod = resource_1.isPod;
exports.isDeployment = resource_1.isDeployment;
exports.isKubeItems = resource_1.isKubeItems;
exports.isJob = resource_1.isJob;
var traffic_light_1 = require("./lib/model/traffic-light");
exports.TrafficLight = traffic_light_1.default;
var apiVersion_1 = require("./controller/kubectl/apiVersion");
exports.apiVersion = apiVersion_1.default;
var exec_1 = require("./controller/kubectl/exec");
exports.doExecWithPty = exec_1.doExecWithPty;
exports.doExecWithStdout = exec_1.doExecWithStdout;
exports.doExecWithStatus = exec_1.doExecWithStatus;
exports.doExecWithTable = exec_1.doExecWithTable;
var raw_1 = require("./controller/kubectl/raw");
exports.doExecRaw = raw_1.doExecRaw;
exports.doNativeExec = raw_1.doNativeExec;
var command_prefix_1 = require("./controller/command-prefix");
exports.commandPrefix = command_prefix_1.default;
var flags_1 = require("./controller/kubectl/flags");
exports.defaultFlags = flags_1.default;
var options_1 = require("./controller/kubectl/options");
exports.hasLabel = options_1.hasLabel;
exports.getLabel = options_1.getLabel;
exports.getLabelForArgv = options_1.getLabelForArgv;
exports.isHelpRequest = options_1.isHelpRequest;
exports.getNamespace = options_1.getNamespace;
exports.getNamespaceForArgv = options_1.getNamespaceForArgv;
exports.isForAllNamespaces = options_1.isForAllNamespaces;
var name_1 = require("./lib/util/name");
exports.parseName = name_1.default;
var formatTable_1 = require("./lib/view/formatTable");
exports.formatTable = formatTable_1.formatTable;
exports.preprocessTable = formatTable_1.preprocessTable;
var help_1 = require("./lib/util/help");
exports.renderHelp = help_1.renderHelp;
var util_1 = require("./lib/util/util");
exports.getCommandFromArgs = util_1.getCommandFromArgs;
var fetch_file_1 = require("./lib/util/fetch-file");
exports.fetchFileString = fetch_file_1.fetchFileString;
var fqn_1 = require("./controller/kubectl/fqn");
exports.fqnOf = fqn_1.fqnOf;
exports.fqn = fqn_1.fqn;
var get_1 = require("./controller/kubectl/get");
exports.doGet = get_1.doGet;
var run_1 = require("./controller/kubectl/run");
exports.doRun = run_1.doRun;
var create_1 = require("./controller/kubectl/create");
exports.doCreate = create_1.doCreate;
var delete_1 = require("./controller/kubectl/delete");
exports.doDelete = delete_1.doDelete;
var describe_1 = require("./controller/kubectl/describe");
exports.doDescribe = describe_1.doDescribe;
