// @kui-shell/plugin-kubeui/dist/lib/view/css-for-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const traffic_light_1 = require("../model/traffic-light");
exports.default = {
    UNKNOWN: '',
    DEPLOYED: traffic_light_1.default.Green,
    DELETED: '',
    SUPERSEDED: traffic_light_1.default.Yellow,
    FAILED: traffic_light_1.default.Red,
    DELETING: traffic_light_1.default.Yellow,
    'Init:0/1': traffic_light_1.default.Yellow,
    PodScheduled: traffic_light_1.default.Yellow,
    PodInitializing: traffic_light_1.default.Yellow,
    Initialized: traffic_light_1.default.Yellow,
    Terminating: traffic_light_1.default.Yellow,
    CrashLoopBackOff: traffic_light_1.default.Red,
    Failed: traffic_light_1.default.Red,
    Running: traffic_light_1.default.Green,
    Pending: traffic_light_1.default.Yellow,
    Succeeded: traffic_light_1.default.Gray,
    Completed: traffic_light_1.default.Gray,
    Unknown: '',
    Ready: traffic_light_1.default.Green,
    ProvisionedSuccessfully: traffic_light_1.default.Green,
    Active: traffic_light_1.default.Green,
    Online: traffic_light_1.default.Green,
    NodeReady: traffic_light_1.default.Green,
    Pulled: traffic_light_1.default.Green,
    Rebooted: traffic_light_1.default.Green,
    Started: traffic_light_1.default.Green,
    Created: traffic_light_1.default.Green,
    Scheduled: traffic_light_1.default.Green,
    SuccessfulCreate: traffic_light_1.default.Green,
    SuccessfulMountVol: traffic_light_1.default.Green,
    ContainerCreating: traffic_light_1.default.Yellow,
    Starting: traffic_light_1.default.Yellow,
    NodeNotReady: traffic_light_1.default.Yellow,
    Killing: traffic_light_1.default.Yellow,
    Deleting: traffic_light_1.default.Yellow,
    Pulling: traffic_light_1.default.Yellow,
    BackOff: traffic_light_1.default.Yellow,
    Unhealthy: traffic_light_1.default.Red,
    FailedScheduling: traffic_light_1.default.Red,
    FailedKillPod: traffic_light_1.default.Red
};
