// @kui-shell/plugin-logui/mdist/modes/show-logs.js
import { i18n } from '@kui-shell/core';
import { isJob, isDeployment, isPod } from '@kui-shell/plugin-kubeui';
const strings = i18n('plugin-logui');
function deploymentLogs(tab, resource) {
    return `kubectl logs deployment/${resource.metadata.name} -n ${resource.metadata.namespace} --all-containers --tail 20`;
}
function jobLogs(tab, resource) {
    return `kubectl logs job/${resource.metadata.name} -n ${resource.metadata.namespace} --tail 20`;
}
function podLogs(tab, resource) {
    return `kubectl logs ${resource.metadata.name} -n ${resource.metadata.namespace} --all-containers --tail 20`;
}
function hasLogs(resource) {
    return isPod(resource) || isDeployment(resource) || isJob(resource);
}
const renderLogs = (tab, resource) => {
    if (isDeployment(resource)) {
        return deploymentLogs(tab, resource);
    }
    else if (isPod(resource)) {
        return podLogs(tab, resource);
    }
    else if (isJob(resource)) {
        return jobLogs(tab, resource);
    }
    else {
        return '';
    }
};
export default {
    when: hasLogs,
    mode: {
        mode: 'logs',
        label: strings('Show Logs'),
        command: renderLogs,
        kind: 'drilldown'
    }
};
