// @kui-shell/plugin-search/dist/views/modes/yaml.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.editSpec = editSpec;
exports.yamlTab = void 0;

var jsYaml = _interopRequireWildcard(require("js-yaml"));

var _core = require("@kui-shell/core");

var _i18n = _interopRequireDefault(require("../../util/i18n"));

var _HTTPClient = _interopRequireDefault(require("../../controller/HTTPClient"));

var _searchQueries = require("../../definitions/search-queries");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

function editSpec(cmd, resource, data) {
  return {
    readOnly: false,
    clearable: false,
    save: {
      label: (0, _i18n.default)('sidecar.yaml.edit.apply'),
      onSave: updated => __awaiter(this, void 0, void 0, function* () {
        (0, _HTTPClient.default)('post', 'console', (0, _searchQueries.UPDATE_RESOURCE)(jsYaml.load(updated), data)).then(() => {
          (0, _core.getCurrentTab)().REPL.pexec(cmd);
        }).catch(err => {
          console.debug(err);
        });
        return {
          noToolbarUpdate: true
        };
      })
    },
    revert: {
      label: (0, _i18n.default)('validation.revert'),
      onRevert: () => jsYaml.safeDump(resource)
    }
  };
}

const yamlTab = (resource, data, cmd) => {
  const spec = editSpec(cmd, resource, data.items[0]);
  return {
    mode: 'yaml',
    label: 'YAML',
    order: 2,
    content: jsYaml.safeDump(resource),
    contentType: 'yaml',
    spec
  };
};

exports.yamlTab = yamlTab;
