// @kui-shell/plugin-kubeui/mdist/lib/view/css-for-value.js
import TrafficLight from '../model/traffic-light';
export default {
    UNKNOWN: '',
    DEPLOYED: TrafficLight.Green,
    DELETED: '',
    SUPERSEDED: TrafficLight.Yellow,
    FAILED: TrafficLight.Red,
    DELETING: TrafficLight.Yellow,
    'Init:0/1': TrafficLight.Yellow,
    PodScheduled: TrafficLight.Yellow,
    PodInitializing: TrafficLight.Yellow,
    Initialized: TrafficLight.Yellow,
    Terminating: TrafficLight.Yellow,
    CrashLoopBackOff: TrafficLight.Red,
    Failed: TrafficLight.Red,
    Running: TrafficLight.Green,
    Pending: TrafficLight.Yellow,
    Succeeded: TrafficLight.Gray,
    Completed: TrafficLight.Gray,
    Unknown: '',
    Ready: TrafficLight.Green,
    ProvisionedSuccessfully: TrafficLight.Green,
    Active: TrafficLight.Green,
    Online: TrafficLight.Green,
    NodeReady: TrafficLight.Green,
    Pulled: TrafficLight.Green,
    Rebooted: TrafficLight.Green,
    Started: TrafficLight.Green,
    Created: TrafficLight.Green,
    Scheduled: TrafficLight.Green,
    SuccessfulCreate: TrafficLight.Green,
    SuccessfulMountVol: TrafficLight.Green,
    ContainerCreating: TrafficLight.Yellow,
    Starting: TrafficLight.Yellow,
    NodeNotReady: TrafficLight.Yellow,
    Killing: TrafficLight.Yellow,
    Deleting: TrafficLight.Yellow,
    Pulling: TrafficLight.Yellow,
    BackOff: TrafficLight.Yellow,
    Unhealthy: TrafficLight.Red,
    FailedScheduling: TrafficLight.Red,
    FailedKillPod: TrafficLight.Red
};
