// @kui-shell/plugin-search/mdist/src-web/controller/helpfiles/searchhelp.js
"use strict";
/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
exports.__esModule = true;
var i18n_1 = require("../../util/i18n");
var sections = [{
        title: i18n_1["default"]('validation.definition.flag'),
        rows: [
            {
                name: '-i',
                docs: i18n_1["default"]('searchhelp.search.install.docs')
            },
            {
                name: '--save',
                docs: i18n_1["default"]('searchhelp.search.save.docs'),
                noclick: true
            },
        ]
    }];
var detailedExample = [
    {
        command: "search <" + i18n_1["default"]('validation.definition.value') + ">",
        docs: i18n_1["default"]('searchhelp.search.value.example')
    },
    {
        command: "search <" + i18n_1["default"]('validation.definition.field') + ">:<" + i18n_1["default"]('validation.definition.value') + ">",
        docs: i18n_1["default"]('searchhelp.search.field.example')
    },
    {
        command: "search <" + i18n_1["default"]('validation.definition.field') + ">:<" + i18n_1["default"]('validation.definition.value') + "> --save",
        docs: i18n_1["default"]('searchhelp.search.save.example')
    },
    {
        command: "search " + i18n_1["default"]('validation.definition.summary') + " <" + i18n_1["default"]('validation.definition.field') + ">:<" + i18n_1["default"]('validation.definition.value') + ">",
        docs: i18n_1["default"]('searchhelp.search.summary.example')
    },
];
/**
 * Usage model for the search plugin
 *
 */
exports.toplevel = {
    breadcrumb: 'search',
    command: 'search',
    title: i18n_1["default"]('searchhelp.title'),
    header: i18n_1["default"]('searchhelp.header'),
    detailedExample: detailedExample,
    example: "search [" + i18n_1["default"]('validation.definition.flag') + "][" + i18n_1["default"]('validation.definition.option') + "]",
    nRowsInViewport: 4,
    available: [
        {
            docs: i18n_1["default"]('searchhelp.title'),
            dir: true,
            commandPrefix: 'search -h'
        },
    ],
    sections: sections,
    alias: 's',
    related: ['savedsearches']
};
