// @kui-shell/plugin-search/mdist/src-web/controller/helpfiles/searchhelp.js
"use strict";
/*******************************************************************************
* Licensed Materials - Property of IBM
* (c) Copyright IBM Corporation 2019. All Rights Reserved.
*
* Note to U.S. Government Users Restricted Rights:
* Use, duplication or disclosure restricted by GSA ADP Schedule
* Contract with IBM Corp.
*******************************************************************************/
exports.__esModule = true;
var i18n_1 = require("../../util/i18n");
/** how many options to show at a time, i.e. before scrolling */
var nRowsInViewport = 1;
/** list of related commands */
var all = [
    'savedsearches',
];
var allExcept = function (cmd) { return all.filter(function (_) { return _ !== cmd; }); };
/* the breadcrumb chain */
var parents = ['search'];
var header = {
    search: i18n_1["default"]('searchhelp.title'),
    savedsearches: i18n_1["default"]('savedsearchhelp.title')
};
/**
 * Usage model for the search plugin
 *
 */
exports.toplevel = {
    breadcrumb: 'search',
    command: 'search',
    title: i18n_1["default"]('searchhelp.title'),
    header: i18n_1["default"]('searchhelp.header') + ".",
    example: "search <" + i18n_1["default"]('validation.definition.value') + ">\nsearch <" + i18n_1["default"]('validation.definition.field') + ">:<" + i18n_1["default"]('validation.definition.value') + ">\nsearch " + i18n_1["default"]('validation.definition.summary') + " <" + i18n_1["default"]('validation.definition.kind') + "> <" + i18n_1["default"]('validation.definition.resource') + ">",
    nRowsInViewport: 4,
    available: [
        {
            docs: header.search,
            dir: true,
            commandPrefix: 'search -h'
        },
    ],
    related: ['savedsearches']
};
