// @kui-shell/core/dist/models/execOptions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.hasLanguage = hasLanguage;
exports.withLanguage = withLanguage;
exports.DefaultExecOptionsForTab = exports.DefaultExecOptions = void 0;

var _command = require("./command");

/*
 * Copyright 2017-18 IBM Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function hasLanguage(execOptions) {
  return execOptions.language !== undefined;
}

function withLanguage(execOptions) {
  if (hasLanguage(execOptions)) {
    return execOptions;
  } else {
    return Object.assign({}, execOptions, {
      language: typeof navigator !== 'undefined' && navigator.language
    });
  }
}

class DefaultExecOptions {
  constructor(type = _command.ExecType.TopLevel) {
    this.type = type;
    this.language = typeof navigator !== 'undefined' && navigator.language;
  }

}

exports.DefaultExecOptions = DefaultExecOptions;

class DefaultExecOptionsForTab extends DefaultExecOptions {
  constructor(tab, block) {
    super();
    this.tab = tab;
    this.block = block;
  }

}

exports.DefaultExecOptionsForTab = DefaultExecOptionsForTab;
