// @kui-shell/plugin-kubectl/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "isKubeResource", {
  enumerable: true,
  get: function () {
    return _resource.isKubeResource;
  }
});
Object.defineProperty(exports, "isCrudableKubeResource", {
  enumerable: true,
  get: function () {
    return _resource.isCrudableKubeResource;
  }
});
Object.defineProperty(exports, "isPod", {
  enumerable: true,
  get: function () {
    return _resource.isPod;
  }
});
Object.defineProperty(exports, "isDeployment", {
  enumerable: true,
  get: function () {
    return _resource.isDeployment;
  }
});
Object.defineProperty(exports, "isKubeItems", {
  enumerable: true,
  get: function () {
    return _resource.isKubeItems;
  }
});
Object.defineProperty(exports, "isKubeItemsOfKind", {
  enumerable: true,
  get: function () {
    return _resource.isKubeItemsOfKind;
  }
});
Object.defineProperty(exports, "isJob", {
  enumerable: true,
  get: function () {
    return _resource.isJob;
  }
});
Object.defineProperty(exports, "TrafficLight", {
  enumerable: true,
  get: function () {
    return _trafficLight.default;
  }
});
Object.defineProperty(exports, "apiVersion", {
  enumerable: true,
  get: function () {
    return _apiVersion.default;
  }
});
Object.defineProperty(exports, "doExecWithMarkdown", {
  enumerable: true,
  get: function () {
    return _execToMarkdown.default;
  }
});
Object.defineProperty(exports, "doExecWithStdoutViaPty", {
  enumerable: true,
  get: function () {
    return _exec.doExecWithStdoutViaPty;
  }
});
Object.defineProperty(exports, "doExecWithPty", {
  enumerable: true,
  get: function () {
    return _exec.doExecWithPty;
  }
});
Object.defineProperty(exports, "doExecWithStdout", {
  enumerable: true,
  get: function () {
    return _exec.doExecWithStdout;
  }
});
Object.defineProperty(exports, "doExecWithStatus", {
  enumerable: true,
  get: function () {
    return _exec.doExecWithStatus;
  }
});
Object.defineProperty(exports, "doExecWithRadioTable", {
  enumerable: true,
  get: function () {
    return _exec.doExecWithRadioTable;
  }
});
Object.defineProperty(exports, "doExecWithTable", {
  enumerable: true,
  get: function () {
    return _exec.doExecWithTable;
  }
});
Object.defineProperty(exports, "doExecRaw", {
  enumerable: true,
  get: function () {
    return _raw.doExecRaw;
  }
});
Object.defineProperty(exports, "doNativeExec", {
  enumerable: true,
  get: function () {
    return _raw.doNativeExec;
  }
});
Object.defineProperty(exports, "commandPrefix", {
  enumerable: true,
  get: function () {
    return _commandPrefix.default;
  }
});
Object.defineProperty(exports, "defaultFlags", {
  enumerable: true,
  get: function () {
    return _flags.default;
  }
});
Object.defineProperty(exports, "crudFlags", {
  enumerable: true,
  get: function () {
    return _flags.crudFlags;
  }
});
Object.defineProperty(exports, "flags", {
  enumerable: true,
  get: function () {
    return _flags.flags;
  }
});
Object.defineProperty(exports, "getCurrentContext", {
  enumerable: true,
  get: function () {
    return _contexts.getCurrentContext;
  }
});
Object.defineProperty(exports, "getCurrentContextName", {
  enumerable: true,
  get: function () {
    return _contexts.getCurrentContextName;
  }
});
Object.defineProperty(exports, "getCurrentDefaultNamespace", {
  enumerable: true,
  get: function () {
    return _contexts.getCurrentDefaultNamespace;
  }
});
Object.defineProperty(exports, "hasLabel", {
  enumerable: true,
  get: function () {
    return _options.hasLabel;
  }
});
Object.defineProperty(exports, "getLabel", {
  enumerable: true,
  get: function () {
    return _options.getLabel;
  }
});
Object.defineProperty(exports, "getLabelForArgv", {
  enumerable: true,
  get: function () {
    return _options.getLabelForArgv;
  }
});
Object.defineProperty(exports, "isHelpRequest", {
  enumerable: true,
  get: function () {
    return _options.isHelpRequest;
  }
});
Object.defineProperty(exports, "getContainer", {
  enumerable: true,
  get: function () {
    return _options.getContainer;
  }
});
Object.defineProperty(exports, "getNamespace", {
  enumerable: true,
  get: function () {
    return _options.getNamespace;
  }
});
Object.defineProperty(exports, "isForAllNamespaces", {
  enumerable: true,
  get: function () {
    return _options.isForAllNamespaces;
  }
});
Object.defineProperty(exports, "parseName", {
  enumerable: true,
  get: function () {
    return _name.default;
  }
});
Object.defineProperty(exports, "stringToTable", {
  enumerable: true,
  get: function () {
    return _formatTable.stringToTable;
  }
});
Object.defineProperty(exports, "formatTable", {
  enumerable: true,
  get: function () {
    return _formatTable.formatTable;
  }
});
Object.defineProperty(exports, "preprocessTable", {
  enumerable: true,
  get: function () {
    return _formatTable.preprocessTable;
  }
});
Object.defineProperty(exports, "isUsage", {
  enumerable: true,
  get: function () {
    return _help.isUsage;
  }
});
Object.defineProperty(exports, "doHelp", {
  enumerable: true,
  get: function () {
    return _help.doHelp;
  }
});
Object.defineProperty(exports, "getCommandFromArgs", {
  enumerable: true,
  get: function () {
    return _util.getCommandFromArgs;
  }
});
Object.defineProperty(exports, "fetchFileString", {
  enumerable: true,
  get: function () {
    return _fetchFile.fetchFileString;
  }
});
Object.defineProperty(exports, "fqnOf", {
  enumerable: true,
  get: function () {
    return _fqn.fqnOf;
  }
});
Object.defineProperty(exports, "fqn", {
  enumerable: true,
  get: function () {
    return _fqn.fqn;
  }
});
Object.defineProperty(exports, "registerEdit", {
  enumerable: true,
  get: function () {
    return _edit.register;
  }
});
Object.defineProperty(exports, "doGet", {
  enumerable: true,
  get: function () {
    return _get.doGet;
  }
});
Object.defineProperty(exports, "getter", {
  enumerable: true,
  get: function () {
    return _get.getter;
  }
});
Object.defineProperty(exports, "getTransformer", {
  enumerable: true,
  get: function () {
    return _get.viewTransformer;
  }
});
Object.defineProperty(exports, "doRun", {
  enumerable: true,
  get: function () {
    return _run.doRun;
  }
});
Object.defineProperty(exports, "doCreate", {
  enumerable: true,
  get: function () {
    return _create.doCreate;
  }
});
Object.defineProperty(exports, "doDelete", {
  enumerable: true,
  get: function () {
    return _delete.doDelete;
  }
});
Object.defineProperty(exports, "describer", {
  enumerable: true,
  get: function () {
    return _describe.describer;
  }
});
Object.defineProperty(exports, "registerConfig", {
  enumerable: true,
  get: function () {
    return _config.register;
  }
});
Object.defineProperty(exports, "onKubectlConfigChangeEvents", {
  enumerable: true,
  get: function () {
    return _config.onKubectlConfigChangeEvents;
  }
});
Object.defineProperty(exports, "offKubectlConfigChangeEvents", {
  enumerable: true,
  get: function () {
    return _config.offKubectlConfigChangeEvents;
  }
});
Object.defineProperty(exports, "registerApplySubcommands", {
  enumerable: true,
  get: function () {
    return _applySubcommands.registerApplySubcommands;
  }
});
Object.defineProperty(exports, "getNamespacesTransformer", {
  enumerable: true,
  get: function () {
    return _getNamespaces.viewTransformer;
  }
});
Object.defineProperty(exports, "Parser", {
  enumerable: true,
  get: function () {
    return _parse.default;
  }
});

var _resource = require("./lib/model/resource");

var _trafficLight = _interopRequireDefault(require("./lib/model/traffic-light"));

var _apiVersion = _interopRequireDefault(require("./controller/kubectl/apiVersion"));

var _execToMarkdown = _interopRequireDefault(require("./controller/kubectl/exec-to-markdown"));

var _exec = require("./controller/kubectl/exec");

var _raw = require("./controller/kubectl/raw");

var _commandPrefix = _interopRequireDefault(require("./controller/command-prefix"));

var _flags = _interopRequireWildcard(require("./controller/kubectl/flags"));

var _contexts = require("./controller/kubectl/contexts");

var _options = require("./controller/kubectl/options");

var _name = _interopRequireDefault(require("./lib/util/name"));

var _formatTable = require("./lib/view/formatTable");

var _help = require("./lib/util/help");

var _util = require("./lib/util/util");

var _fetchFile = require("./lib/util/fetch-file");

var _fqn = require("./controller/kubectl/fqn");

var _edit = require("./controller/kubectl/edit");

var _get = require("./controller/kubectl/get");

var _run = require("./controller/kubectl/run");

var _create = require("./controller/kubectl/create");

var _delete = require("./controller/kubectl/delete");

var _describe = require("./controller/kubectl/describe");

var _config = require("./controller/kubectl/config");

var _applySubcommands = require("./controller/kubectl/apply-subcommands");

var _getNamespaces = require("./controller/kubectl/get-namespaces");

var _parse = _interopRequireDefault(require("./lib/util/parse"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
